Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch

```diff
@@ -0,0 +1,165 @@
++                                    event.setCancelled(true);
++                                }
++                                world.getServer().getPluginManager().callEvent(event);
++                            }
++                        }
++                        //world.getServer().getPluginManager().callEvent(event);
++                        // Paper end
+                         if (event.isCancelled()) {
+                             return;
+                         }
+diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
+index 1f5de89..87648d1 100644
+--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
++++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
+@@ -263,7 +263,18 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+                         }
+ 
+                         InventoryMoveItemEvent event = new InventoryMoveItemEvent(this.getOwner().getInventory(), oitemstack.clone(), destinationInventory, true);
+-                        this.getWorld().getServer().getPluginManager().callEvent(event);
++                        // Paper start - call event as cancelled when inventory wont accept item in Vanilla
++                        if (world.paperConfig.callMoveItemEvent) {
++                            boolean acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
++                            if (world.paperConfig.callMoveItemEventWithoutMove || acceptsItem) {
++                                if (world.paperConfig.callMoveEventCancelled && !acceptsItem) {
++                                    event.setCancelled(true);
++                                }
++                                this.getWorld().getServer().getPluginManager().callEvent(event);
++                            }
++                        }
++                        // this.getWorld().getServer().getPluginManager().callEvent(event);
++                        // Paper end
+                         if (event.isCancelled()) {
+                             this.setItem(i, itemstack);
+                             this.setCooldown(world.spigotConfig.hopperTransfer); // Spigot
+@@ -420,8 +431,18 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+             }
+ 
+             InventoryMoveItemEvent event = new InventoryMoveItemEvent(sourceInventory, oitemstack.clone(), ihopper.getOwner().getInventory(), false);
+-
+-            ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++            // Paper start - call event as cancelled when inventory wont accept item in Vanilla
++            if (ihopper.getWorld().paperConfig.callMoveItemEvent) {
++                boolean acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
```

Comment:
It's not as easy as that. The check needs to be also made when callMoveEventCancelled(withoutMove, maybe the variable name should be more descriptive) is true.

The code would have to be something like this:
```
boolean acceptsItem = true;
if (!world.paperConfig.callMoveItemEventWithoutMove || world.paperConfig.callMoveEventCancelledWithoutMove) {
    acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
}
if (world.paperConfig.callMoveItemEventWithoutMove || acceptsItem) {
    if (world.paperConfig.callMoveEventCancelledWithoutMove && !acceptsItem) {
        event.setCancelled(true);
    }
    this.getWorld().getServer().getPluginManager().callEvent(event);
}
```
The acceptsItem variable assignment could be compressed to `boolean acceptsItem = (!world.paperConfig.callMoveItemEventWithoutMove || world.paperConfig.callMoveEventCancelledWithoutMove && willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);` but that's kinda hard to read. But I don't see how you could get the willAcceptItem into the other statement. Not sure if it is worth to expand the calculation of how the event should be called to get rid of the many nested if statements. If so it might be best to move it into its own method?